java/ql/src/experimental/semmle/code/java/security/DecompressionBomb.qll

```diff
@@ -0,0 +1,713 @@
+    }
+
+    /**
+     * A method Access as a sink which responsible for reading bytes
+     */
+    MethodAccess getAByteRead() { result = this }
+
+    // while ((readLen = zipInputStream.read(readBuffer)) != -1) {
+    //   totallRead += readLen;
+    //   if (totallRead > 1024 * 1024 * 4) {
+    //     System.out.println("potential Bomb");
+    //     break;
+    //   }
+    //   outputStream.write(readBuffer, 0, readLen);
+    // }
+    // TODO: I don't know why we can't reach totallRead with Local Tainting
+    // the same behaviour exists in golang
+    predicate isControlledRead() {
+      exists(ComparisonExpr i |
+        TaintTracking::localExprTaint([this, this.getArgument(2)], i.getAChildExpr*())
+      )
+    }
+  }
+
+  class Sink extends DecompressionBomb::Sink {
+    override predicate sink(DataFlow::Node sink, DataFlow::FlowState state) {
+      sink.asExpr() = any(ReadInputStreamCall r).getAByteRead() and state = "Zip4j"
+    }
+  }
+
+  /**
+   * Gets `n1` and `n2` which `ZipInputStream n2 = new ZipInputStream(n1)` or `n2 = zipInputStream.read(n1)` or `n1.read(n2)`,
+   * second one is added because of sanitizer, we want to compare return value of each `read` or similar method
+   * that whether there is a flow to a comparison between total read of decompressed stream and a constant value
+   */
+  private class InputStreamAdditionalTaintStep extends DecompressionBomb::AdditionalStep {
+    override predicate step(
+      DataFlow::Node n1, DataFlow::FlowState stateFrom, DataFlow::Node n2,
+      DataFlow::FlowState stateTo
+    ) {
+      exists(Call call |
+        // Constructors
+        call.getCallee().getDeclaringType() = any(TypeZipInputStream t) and
+        call.getArgument(0) = n1.asExpr() and
+        call = n2.asExpr()
+        or
+        // Method calls
+        call.(MethodAccess).getReceiverType() = any(TypeZipInputStream t) and
+        call.getCallee().hasName(["read", "readNBytes", "readAllBytes"]) and
+        call.getQualifier() = n1.asExpr() and
+        call = n2.asExpr()
+      ) and
+      stateFrom = "Zip4j" and
+      stateTo = "Zip4j"
+    }
+  }
+}
+
+/**
+ * Providing sinks that can be related to reading uncontrolled buffer and bytes for `org.apache.commons.io` package
+ */
+module CommonsIO {
+  /**
+   * The Access to Methods which work with byes and inputStreams and buffers
+   */
+  class IOUtils extends MethodAccess {
+    IOUtils() {
+      this.getMethod()
+          .hasName([
+              "copy", "copyLarge", "read", "readFully", "readLines", "toBufferedInputStream",
+              "toByteArray", "toCharArray", "toString", "buffer"
+            ]) and
+      this.getMethod().getDeclaringType().hasQualifiedName("org.apache.commons.io", "IOUtils")
+    }
+  }
+
+  class Sink extends DecompressionBomb::Sink {
+    override predicate sink(DataFlow::Node sink, DataFlow::FlowState state) {
+      sink.asExpr() = any(IOUtils r).getArgument(0) and
```

Comment:
This sink is too noisy as it reports the IOUtils calls on any java.io.InputStream. Since the remote flow sources are affixed the "decompression" flow states and they flow them straight to the sink without flowing through any decompression inputstreams.

For example, this is reported:

```
xmlResult = IOUtils.toString(request.getInputStream(), request.getCharacterEncoding());
```

I think you may want to NOT assign a flow state at the source and then affix it at the relevant decompression operations such as the decompression input stream constructors. At the sink, if the taint has any of these flow states, that would mean that it went through some decompression.